src/Test/Utilities/Portable/Metadata/MetadataReaderUtils.cs

```diff
@@ -265,9 +265,147 @@ public static string Dump(this MetadataReader reader, EntityHandle handle)
+            {
+                return elementType + "*";
+            }
+
+            public string GetPrimitiveType(PrimitiveTypeCode typeCode)
+            {
+                return typeCode.ToString();
+            }
+
+            public string GetSZArrayType(string elementType)
+            {
+                return elementType + "[]";
+            }
+
+            public string GetTypeFromDefinition(MetadataReader reader, TypeDefinitionHandle handle, byte rawTypeKind)
+            {
+                var typeDef = reader.GetTypeDefinition(handle);
+                var name = reader.GetString(typeDef.Name);
+                return typeDef.Namespace.IsNil ? name : reader.GetString(typeDef.Namespace) + "." + name;
+            }
+
+            public string GetTypeFromReference(MetadataReader reader, TypeReferenceHandle handle, byte rawTypeKind)
+            {
+                var typeRef = reader.GetTypeReference(handle);
+                var name = reader.GetString(typeRef.Name);
+                return typeRef.Namespace.IsNil ? name : reader.GetString(typeRef.Namespace) + "." + name;
+            }
+
+            public string GetTypeFromSpecification(MetadataReader reader, object genericContext, TypeSpecificationHandle handle, byte rawTypeKind)
+            {
+                var sigReader = reader.GetBlobReader(reader.GetTypeSpecification(handle).Signature);
+                return new SignatureDecoder<string, object>(Instance, reader, genericContext).DecodeType(ref sigReader);
+            }
+        }
+
+        internal static void VerifyPEMetadata(string pePath, string[] types, string[] methods, string[] attributes)
+        {
+            using (var peStream = File.OpenRead(pePath))
+            using (var refPeReader = new PEReader(peStream))
+            {
+                var metadataReader = refPeReader.GetMetadataReader();
+
+                AssertEx.SetEqual(metadataReader.TypeDefinitions.Select(t => metadataReader.Dump(t)), types);
+                AssertEx.SetEqual(metadataReader.MethodDefinitions.Select(t => metadataReader.Dump(t)), methods);
+
+                AssertEx.SetEqual(
+                    metadataReader.CustomAttributes.Select(a => metadataReader.GetCustomAttribute(a).Constructor)
+                        .Select(c => metadataReader.GetMemberReference((MemberReferenceHandle)c).Parent)
+                        .Select(p => metadataReader.GetTypeReference((TypeReferenceHandle)p).Name)
+                        .Select(n => metadataReader.GetString(n)),
+                    attributes);
+            }
+        }
+
+        /// <summary>
+        /// Asserts that all the method bodies in the PE are empty or `throw null`, if expectEmptyOrThrowNull is true.
+        /// Asserts that none of the methods bodies are empty or `throw null`, if expectEmptyOrThrowNull is false.
+        /// </summary>
+        internal static void VerifyMethodBodies(ImmutableArray<byte> peImage, bool expectEmptyOrThrowNull)
```

Comment:
>VerifyMethodBodies [](start = 29, length = 18)

This is an odd helper. Perhaps change to VerifyMethodBodies(peImage, Action<byte[]> ilVerifier)